performance_test/src/communication_abstractions/ros2_callback_communicator.hpp

```diff
@@ -21,164 +21,44 @@
-    } else if (m_qos.durability == QOSAbstraction::Durability::TRANSIENT_LOCAL) {
-      ros_qos.durability = rmw_qos_durability_policy_t::RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
-    } else {
-      throw std::runtime_error("Unsupported QOS!");
-    }
-
-
-    if (m_qos.history_kind == QOSAbstraction::HistoryKind::KEEP_ALL) {
-      ros_qos.history = rmw_qos_history_policy_t::RMW_QOS_POLICY_HISTORY_KEEP_ALL;
-    } else if (m_qos.history_kind == QOSAbstraction::HistoryKind::KEEP_LAST) {
-      ros_qos.history = rmw_qos_history_policy_t::RMW_QOS_POLICY_HISTORY_KEEP_LAST;
-    } else {
-      throw std::runtime_error("Unsupported QOS!");
-    }
-    ros_qos.depth = m_qos.history_depth;
-
-    return ros_qos;
-  }
-
-private:
-  const QOSAbstraction m_qos;
-};
-
 /// Communication plugin for ROS 2 using ROS 2 callbacks for the subscription side.
 template<class Topic>
-class ROS2CallbackCommunicator : public Communicator
+class ROS2CallbackCommunicator : public ROS2Communicator<Topic>
 {
 public:
   /// The data type to publish and subscribe to.
-  using DataType = typename Topic::RosType;
+  using DataType = typename ROS2Communicator<Topic>::DataType;
 
   /// Constructor which takes a reference \param lock to the lock to use.
   explicit ROS2CallbackCommunicator(SpinLock & lock)
-  : Communicator(lock),
-    m_node(ResourceManager::get().ros2_node()),
-    m_publisher(nullptr),
+  : ROS2Communicator<Topic>(lock),
     m_subscription(nullptr)
   {
-    m_executor.add_node(m_node);
-  }
-
-  /**
-   * \brief Publishes the provided data.
-   *
-   *  The first time this function is called it also creates the publisher.
-   *  Further it updates all internal counters while running.
-   *
-   * \param data The data to publish.
-   * \param time The time to fill into the data field.
-   */
-  void publish(DataType & data, const std::chrono::nanoseconds time)
-  {
-    if (!m_publisher) {
-      auto qos = ROS2QOSAdapter(m_ec.qos()).get();
-      // Workaround for bug where RMW/FastRPS keeps history of volatile publisher.
-      if (qos.durability == rmw_qos_durability_policy_t::RMW_QOS_POLICY_DURABILITY_VOLATILE) {
-        qos.history = rmw_qos_history_policy_t::RMW_QOS_POLICY_HISTORY_KEEP_LAST;
-        qos.depth = std::size_t(10);  // Setting depth to 10 as a safety net.
-      }
-      // End of workaround.
-
-      m_publisher = m_node->create_publisher<DataType>(
-        Topic::topic_name() + m_ec.pub_topic_postfix(), qos);
-    }
-    lock();
-    data.time = time.count();
-    data.id = next_sample_id();
-    increment_sent();  // We increment before publishing so we don't have to lock twice.
-    unlock();
-    m_publisher->publish(data);
+    m_executor.add_node(this->m_node);
   }
 
   /// Reads received data from ROS 2 using callbacks
-  void update_subscription()
+  void update_subscription() override
   {
+    auto ros2QOSAdapter = this->m_ROS2QOSAdapter;
```

Comment:
Why do you create this temporary here?

Reply:
Yes I can remove this. There is no need to do this temporary assignment here.